src/metric/xentropy_metric.hpp

```diff
@@ -0,0 +1,384 @@
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          double hhat = 0;
+          objective->ConvertOutput(&score[i], &hhat); // NOTE: this only works if objective = "xentlambda"
+          sum_loss += XentLoss1(label_[i], 1.0f, hhat);
+        }
+      } else {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          double hhat = 0;
+          objective->ConvertOutput(&score[i], &hhat); // NOTE: this only works if objective = "xentlambda"
+          sum_loss += XentLoss1(label_[i], weights_[i], hhat);
+        }
+      }
+    }
+    return std::vector<double>(1, sum_loss / static_cast<double>(num_data_));
+  }
+
+  const std::vector<std::string>& GetName() const override {
+    return name_;
+  }
+
+  double factor_to_bigger_better() const override {
+    return -1.0f;
+  }
+
+private:
+  /*! \brief Number of data points */
+  data_size_t num_data_;
+  /*! \brief Pointer to label */
+  const float* label_;
+  /*! \brief Pointer to weights */
+  const float* weights_;
+  /*! \brief Name of this metric */
+  std::vector<std::string> name_;
+};
+
+//
+// KullbackLeiblerDivergence : "kldiv" : (optional) weights are used linearly
+//
+class KullbackLeiblerDivergence : public Metric {
+public:
+  explicit KullbackLeiblerDivergence(const MetricConfig&) {}
+  virtual ~KullbackLeiblerDivergence() {}
+
+  void Init(const Metadata& metadata, data_size_t num_data) override {
+    name_.emplace_back("kldiv");
+    num_data_ = num_data;
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+
+    // ensure that labels are in interval [0, 1], interval ends included
+    if (label_ == nullptr) {
+       Log::Fatal("[%s]: label nullptr", __func__);
+    }
+    for (data_size_t i = 0; i < num_data; ++i) {
+      if (label_[i] < 0.0f || label_[i] > 1.0f) {
+        Log::Fatal("[%s]: metric does not tolerate label [#%i] outside [0, 1]", __func__, i);
+      }
+    }
+    Log::Info("[%s:%s]: labels passed interval [0, 1] check (metric)",  GetName()[0].c_str(), __func__);
+
+    if (weights_ == nullptr) {
+      sum_weights_ = static_cast<double>(num_data_);
+    } else {
+      float min_weight = weights_[0];
+      sum_weights_ = 0.0f;
+      for (data_size_t i = 0; i < num_data; ++i) {
+        sum_weights_ += weights_[i];
+        if (weights_[i] < min_weight) min_weight = weights_[i];
+      }
+      // warn if there is at least one negative weight
+      if (min_weight < 0.0f) {
+        Log::Warning("[%s:%s]: minimum weight = %f is negative", GetName()[0].c_str(), __func__, min_weight);
+      }
+    }
+
+    // check weight sum
+    if (sum_weights_ <= 0.0f) {
+      Log::Fatal("[%s]: sum-of-weights = %f is non-positive", __func__, sum_weights_);
+    } else {
+      Log::Info("[%s:%s]: sum-of-weights = %f",  GetName()[0].c_str(), __func__, sum_weights_);
+    }

```

Comment:
There are many duplicated code for the checking the ```label``` and ```weight``` in this file. To avoid that, I think you can define some inline static functions. like ```CheckLabel(const float* label, data_size_t num_data)``` .